File: app/services/user_service.py
from uuid import UUID

from sqlalchemy.orm import Session

from app.repositories.user_repository import UserRepository
from app.core.security import hash_password


class UserService:

    @staticmethod
    def get_all_users(db: Session):
        return UserRepository.get_all(db)

    @staticmethod
    def get_user(db: Session, user_id: UUID):
        return UserRepository.get_by_id(db, user_id)

    @staticmethod
    def update_user(db: Session, user, data):
        if data.username is not None:
            user.username = data.username

        if data.email is not None:
            user.email = data.email

        if data.phone_number is not None:
            user.phone_number = data.phone_number

        if data.password:
            user.password_hash = hash_password(data.password)

        # Role assignment is allowed only through the admin-protected
        # API route in app/api/routes/users.py.
        if data.role is not None:
            user.role = data.role

        return UserRepository.update(db, user)

    @staticmethod
    def delete_user(db: Session, user):
        UserRepository.delete(db, user)

File: frontend/src/services/userService.js
import api from "./api";

// ============================================================
// GET ALL USERS
// ============================================================

export const getUsers = async () => {
    const response = await api.get("/users/");
    return response.data;
};


// ============================================================
// GET USER BY ID
// ============================================================

export const getUserById = async (id) => {
    const response = await api.get(
        `/users/${id}`
    );

    return response.data;
};


// ============================================================
// UPDATE USER
// ============================================================

export const updateUser = async (
    id,
    userData
) => {
    const response = await api.put(
        `/users/${id}`,
        userData
    );

    return response.data;
};


// ============================================================
// DELETE USER
// ============================================================

export const deleteUser = async (id) => {
    const response = await api.delete(
        `/users/${id}`
    );

    return response.data;
};


// ============================================================
// UPDATE USER ROLE
// ============================================================

export const updateUserRole = async (
    userId,
    role
) => {

    if (!userId) {
        throw new Error(
            "User ID is required."
        );
    }

    if (!role) {
        throw new Error(
            "Role is required."
        );
    }

    const response = await api.put(
        `/admin/users/${userId}/role`,
        {
            role: role,
        }
    );

    /*
     * Backend returns:
     *
     * {
     *   message: "...",
     *   user: {
     *      id,
     *      username,
     *      email,
     *      role
     *   }
     * }
     */

    return (
        response.data?.user ||
        response.data
    );
};